### PR TITLE
T/312: Optimized Relation Fixer plugin

### LINKED ISSUE #312: JSDoc: Optimize Relation Fixer plugin
Relation Fixer plugin when run on the sources of CKEditor 5 currently takes up to 10 seconds. There is room for optimizations and I'm going to introduce them.

### PR DESCRIPTION
**Pull request by m-turek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Optimized Relation Fixer plugin. Closes #312.

---

### Additional information

Used DocletCollection to reduce the number of operations on arrays of doclets.

### CODE REVIEW

File: packages/jsdoc-plugins/lib/relation-fixer/addmissingdoclets.js [hunk 1]
```diff
@@ -20,6 +21,13 @@ module.exports = addMissingDoclets;
  */
 function addMissingDoclets( originalDoclets ) {
 	const clonedDoclets = cloneDeep( originalDoclets );
+	const docletCollection = new DocletCollection();
+
+	for ( const doclet of clonedDoclets ) {
+		// group doclets by memberof property
```
Comment (pomek) [inline packages/jsdoc-plugins/lib/relation-fixer/addmissingdoclets.js]: A sentence should start with an uppercase letter and end with a period.

File: packages/jsdoc-plugins/lib/relation-fixer/buildrelations.js [hunk 2]
```diff
@@ -27,15 +28,21 @@ module.exports = addRelationArrays;
  */
 function addRelationArrays( originalDoclets ) {
 	const clonedDoclets = cloneDeep( originalDoclets );
+	const docletCollection = new DocletCollection();
+
+	for ( const doclet of clonedDoclets ) {
+		// group doclets by longname
```
Comment (ma2ciek) [inline packages/jsdoc-plugins/lib/relation-fixer/buildrelations.js]: > A sentence should start with an uppercase letter and end with a period.

Same here.

File: packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js [hunk 3]
```diff
@@ -34,14 +34,14 @@ function getMissingDocletsData( originalDoclets, childDoclet, options ) {
 		clonedDoclet.memberof = childDoclet.longname;
 
 		// Add property `inherited` or `mixed`.
-		const relationProperty = getRelationProperty( originalDoclets, childDoclet, d, options.relation );
+		const relationProperty = getRelationProperty( docletCollection.getAll(), childDoclet, d, options.relation );
```
Comment (ma2ciek) [inline packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js]: Can we pass the whole collection here?

File: packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js [hunk 4]
```diff
@@ -85,7 +85,7 @@ function getDocletsToAdd( allDoclets, childDoclet, options = {} ) {
 				}
 			}
 
-			return d.memberof === longname && isMatchingFilterOptions && !isUnwanted;
+			return isMatchingFilterOptions && !isUnwanted;
 		} );
 
 		docletsToAdd.push( ...toAdd );
```
Comment (ma2ciek) [inline packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js]: You're mixing functional programming with procedural programming. I'd either use `Array.filter()` or push elements to the original array.

E.g. return `docletsToAdd.concat( toAdd )` should be better.

Comment (Reinmar) [reply packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js]: I see that this code wasn't touched in this PR. So, unless this causes some issues, it's beyond the scope of it.

**Comment by ma2ciek** [comment]
I'd change that method: https://github.com/ckeditor/ckeditor5-dev/blob/664fb2d6d4ceab57c3e814f208848dabe6f547e4/packages/jsdoc-plugins/lib/utils/doclet-collection.js#L24-L27 as the `name` has become incorrect. Maybe `category` would fit better?

**Comment by ma2ciek** [comment]
Actually, you could correct more occurrences of the 'record' and 'object' after the previous PR. I didn't notice that.

File: packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js (hunk 3 above)
Comment (m-turek) [reply packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js]: We can, but why? does it matter? (maybe I'm missing something)

**Comment by m-turek** [comment]
> Actually, you could correct more occurrences of the 'record' and 'object' after the previous PR. I didn't notice that.

Is it ok to do it in this PR? or should I make another one for that?

**Comment by ma2ciek** [comment]
> Is it ok to do it in this PR? or should I make another one for that?

I think, that we can do this in that PR as it's related to current changes.

**Comment by ma2ciek** [comment]
Thanks for those changes to `DocletColelction`.

**Pull request merged**